### PR TITLE
Feature/health check

### PR DESCRIPTION
**Pull request by mr-rooftop**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,10 +12,11 @@ Install `docker` and run:
 
 ```
 docker volume create thornode-bot-volume
-docker run -d --env TELEGRAM_BOT_TOKEN=XXX --env DEBUG={True|False} --mount source=thornode-bot-volume,target=/storage block42blockchaincompany/thornode_bot
+docker run -d --env TELEGRAM_BOT_TOKEN=XXX --mount source=thornode-bot-volume,target=/storage block42blockchaincompany/thornode_bot
+docker run -d --name autoheal --restart=always -v /var/run/docker.sock:/var/run/docker.sock willfarrell/autoheal
```
Comment (rekt0x) [inline README.md:16]: Maybe we should use docker-compose to make things easier

File: README.md [hunk 2]
```diff
@@ -103,3 +106,45 @@ docker run --env TELEGRAM_BOT_TOKEN=XXX --mount source=thornode-bot-volume,targe
...
+The script assures that `thornode_bot.py` is periodically updating the `health.check` file.
+
+If the docker health check fails, the docker container is marked as "unhealthy". 
+However, when using docker standalone (without docker compose or docker swarm) this doesn't do anything.
+To restart unhealthy containers, we have to use the autoheal image https://hub.docker.com/r/willfarrell/autoheal/ .
+
+To make sure a potentially unhealthy thorbot_node container is restarted, run the autoheal container alongside the
+thornode_bot container:
+```
+docker run -d --name autoheal --restart=always -v /var/run/docker.sock:/var/run/docker.sock willfarrell/autoheal
```
Comment (rekt0x) [inline README.md:122]: Maybe we should use docker-compose to make things easier

File: healthcheck.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from datetime import datetime, timedelta
+
+HEALTHY = 0
+UNHEALTHY = 1
+
+
+def check_health():
+    return UNHEALTHY
```
Comment (rekt0x) [inline healthcheck.py:8]: I think return statement shouldnt be here

File: Dockerfile [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 
 ADD thornode_bot.py/ /
 ADD requirements.txt/ /
-ADD test/nodeaccounts.json /test
 
 RUN pip install -r requirements.txt
 
-CMD [ "python", "./thornode_bot.py" ]
\ No newline at end of file
+HEALTHCHECK --start-period=10s --interval=30s --retries=1 --timeout=3s CMD [ "python", "./healthcheck.py" ]
```
Comment (rekt0x) [inline Dockerfile:13]: I think i'd go with: `--interval=10s --retries=2`

File: healthcheck.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    timestamp_benchmark = datetime.timestamp(datetime.now() - timedelta(seconds=30))
+    with open('storage/health.check') as healthcheck_file:
+        thornode_bot_timestamp = healthcheck_file.read()
```
Comment (rekt0x) [inline healthcheck.py:11]: Could be a one-liner: last_thornode_bot_timestamp =  open('storage/health.check').read()

File: healthcheck.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    if float(thornode_bot_timestamp) > timestamp_benchmark:
```
Comment (rekt0x) [inline healthcheck.py:13]: I'd change this if-check the following way:

```
# Check if the thornode bot has shown a sign of life within the last 30 seconds
if int(last_thornode_bot_timestamp) >= int(datetime.timestamp(datetime.now() - timedelta(seconds=30))):
...
```

File: healthcheck.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+from datetime import datetime, timedelta
+
+HEALTHY = 0
+UNHEALTHY = 1
+
+
+def check_health():
+    return UNHEALTHY
+    timestamp_benchmark = datetime.timestamp(datetime.now() - timedelta(seconds=30))
```
Comment (rekt0x) [inline healthcheck.py:9]: I'd add this line directly into the if-check below.

File: storage/health.check [hunk 8]
```diff
@@ -0,0 +1 @@
+1587528794.768947
```
Comment (rekt0x) [inline storage/health.check:1]: Exclude that file via .gitignore

File: thornode_bot.py [hunk 9]
```diff
@@ -225,6 +226,15 @@ def check_thornode(context):
         show_action_buttons(context, chat_id=chat_id)
 
 
+def healthy(context):
+    """
+        Write timestamp into health.check file for the health check
+    """
+    timestamp = datetime.timestamp(datetime.now())
+    with open('storage/health.check', 'w') as healthcheck_file:
+        healthcheck_file.write(str(timestamp))
```
Comment (rekt0x) [inline thornode_bot.py:235]:
```
with open('storage/health.check', 'w') as healthcheck_file:
  timestamp = int(datetime.timestamp(datetime.now()))
  healthcheck_file.write(str(timestamp))
```

File: thornode_bot.py [hunk 10]
```diff
@@ -290,10 +300,13 @@ def main():
     # Restart jobs for all users
     chat_ids = dispatcher.user_data.keys()
     for chat_id in chat_ids:
-        dispatcher.job_queue.run_repeating(check_thornode, interval=5, context={
+        dispatcher.job_queue.run_repeating(check_thornode, interval=30, context={
             'chat_id': chat_id, 'user_data': dispatcher.user_data[chat_id]
         })
 
+    # Start job for health check
+    dispatcher.job_queue.run_repeating(healthy, interval=10, context={})
```
Comment (rekt0x) [inline thornode_bot.py:308]: I'd run the job all 5 seconds

**Comment by rekt0x** [comment]
I believe you have not rebased this branch onto master, so there are merge conflicts, so rebase it.

File: README.md (hunk 1 above)
Comment (mr-rooftop) [reply README.md:16]: I thought we should do this in the config file feature, I think there we can use some synergy with docker compose and the config file.

File: README.md (hunk 2 above)
Comment (mr-rooftop) [reply README.md:122]: I think better in config file feature

File: healthcheck.py (hunk 3 above)
Comment (mr-rooftop) [reply healthcheck.py:8]: Sorry for all this unnecessary errors by me..

File: storage/health.check (hunk 8 above)
Comment (mr-rooftop) [reply storage/health.check:1]: I thought it's good to have a sample health.check file if someone wants to understand the README. But I can remove it :+1:

File: README.md (hunk 1 above)
Comment (rekt0x) [reply README.md:16]: Sounds good 👍

File: README.md (hunk 2 above)
Comment (rekt0x) [reply README.md:122]: Sounds good 👍

**Pull request closed without merging**